internal/pkg/agent/vault/aesgcm.go

```diff
@@ -0,0 +1,108 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package vault
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"encoding/hex"
+	"syscall"
+)
+
+type AESKeyType int
+
+const (
+	AES128 AESKeyType = 16
+	AES192 AESKeyType = 24
+	AES256 AESKeyType = 32
+)
+
+// NewKey generates new AES key as bytes
+func NewKey(kt AESKeyType) ([]byte, error) {
+	key := make([]byte, kt)
+	if _, err := rand.Read(key); err != nil {
+		return nil, err
+	}
+	return key, nil
+}
+
+// NewKeyHexString generates new AES key as hex encoded string
+func NewKeyHexString(kt AESKeyType) (string, error) {
+	key, err := NewKey(kt)
+	if err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(key), nil
+}
+
+// Encrypt encrypts the data with AES-GCM
+func Encrypt(key, data []byte) ([]byte, error) {
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+
+	aesGCM, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+
+	nonce := make([]byte, aesGCM.NonceSize())
+	if _, err = rand.Read(nonce); err != nil {
```

Comment:
I think you need to use ReadFull to protect against a short read
https://pkg.go.dev/crypto/cipher@go1.18.1#example-NewGCM-Encrypt

Reply:
if you check implementation and the comment for Read:
```
// Read is a helper function that calls Reader.Read using io.ReadFull.
// On return, n == len(b) if and only if err == nil.
func Read(b []byte) (n int, err error) {
	return io.ReadFull(Reader, b)
}
```

https://github.com/golang/go/blob/master/src/crypto/rand/rand.go